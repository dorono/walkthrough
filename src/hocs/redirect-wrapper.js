import React from 'react';
import {Redirect} from 'react-router-dom';
import {has} from 'lodash';

@withRouter
export const redirectWrapper = (redirectUrl) => (Component) => {
    class RedirectWrapper extends React.Component {
        render() {
            // don't redirect if redirectUrl is specified or
            // redirect is to current path
            if (
                redirectUrl
                && this.props.location.pathname !== redirectUrl
            ) {
                return (
                    <Redirect
                        push
                        to={{
                            pathname: redirectUrl,
                            // TODO: Make the state obj more customized
                            state: {
                                status: this.props.status,
                                appName: this.props.appName,
                            },
                        }}
                    />
                );
            }

            // in case user refreshes error page
            const updatedProps = Object.assign({}, this.props);
            if (
                !has(updatedProps, 'location.state.status')
                || typeof updatedProps.location.state.status !== 'number'
            ) {
                updatedProps.location.state = {
                    status: 500,
                };
            }

            return (
                <Component {...updatedProps} />
            );
        }
    }

    return RedirectWrapper;
};
