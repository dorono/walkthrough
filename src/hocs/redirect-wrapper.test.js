import React from 'react';
import {shallow} from 'enzyme';
import {redirectWrapper} from './redirect-wrapper';

const RedirectWrapper = redirectWrapper()(<h1>Test</h1>);

describe('redirectWrapper', () => {
    it('should render without errors', () => {
        shallow(<RedirectWrapper />);
    });

    it('should match snapshot', () => {
        const redirectWrapper = shallow(<RedirectWrapper />);
        expect(redirectWrapper).toMatchSnapshot();
    });
});
