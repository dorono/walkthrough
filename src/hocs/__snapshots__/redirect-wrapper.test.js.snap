// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redirectWrapper should match snapshot 1`] = `
<Route>
  <Component />
</Route>
`;
